### DISCUSSION TITLE
Adding more nodes and edges to existing database and creating relationships with existing nodes/edges

### DISCUSSION THREAD

**Discussion by kamalikaray** [kuzudb/kuzu#3594] [Q&A] [answered]
Hi,

I have a few questions:
Is there a way in kuzu:
1) Build a connection and add nodes/edges to an existing database
2) Creating relationships with existing nodes or edges in the old db
3) Changing or adding more schema 

Thanks,
Kamalika

**Comment by prrao87** [comment, marked as answer]
Hi @kamalikaray, yes those are possible. I'll try to address each of your questions below:

#### 1. Add nodes/edges to an existing database

This can be done in one of two ways. The first is for bulk import cases, where you want to add a large amount of data at once.

```cypher
// Create node table
CREATE NODE TABLE Person(name STRING, age INT64, PRIMARY KEY (name));
// Run COPY FROM with a subquery
COPY Person FROM (LOAD FROM "person1.csv" RETURN *);
COPY Person FROM (LOAD FROM "person2.csv" RETURN *);
```

This allows you to individually import from different source files, assuming they conform to the schema. It also requires some prior work upfront to ensure that you do not have primary key conflicts in either file, so this approach makes sense when you have more faith in your upstream source of data that it provides data that's not duplicated in terms of primary key identifiers.

If you cannot ensure that the data doesn't have duplicates, you could resort to the `MERGE` keyword [as per the docs](https://docs.kuzudb.com/cypher/data-manipulation-clauses/merge/#merge), which will first match on the primary key identifier specified, and the properties specified by the SET clause will overwrite existing properties in the graph.

#### 2. Creating relationships with existing nodes in the old db

This is simple: you first match on the existing nodes between which you want to create the edge, and then use a merge clause that overwrites the properties of an existing edge (or simply creates a new edge if it doesn't exist). See [the docs](https://docs.kuzudb.com/cypher/data-manipulation-clauses/merge/#merge-existing-relationships) for more on this.

#### 3. Changing or adding more schema

Yes, it's possible to update the schema once a table exists using the `ALTER TABLE` command. You can also add more columns (that represent additional node/relationship properties) to the table without having to drop the entire table. However, changing the data types of a column in place is not possible - you'd have to drop the existing column and add the new column of the desired type. However, this is quite simple (and very fast) if you're using bulk imports and you already have the full table's columns in memory (for e.g., via a Pandas DataFrame), or on disk (e.g., in a CSV or Parquet file). [The docs](https://docs.kuzudb.com/cypher/data-definition/alter/) for `ALTER TABLE` go into more detail on this.

Feel free to mark this as the answer if it addresses your questions :)

**Reply by kamalikaray** [reply]
Hi,
Thanks a lot for the detailed answers. 
Is the command for connecting to the existing database similar to starting a new one?
`db = kuzu.Database("./test")`
`conn = kuzu.Connection(db)`

**Reply by prrao87** [reply]
Yes, that's the right way. The first line creates a new database directory if it doesn't exist, but connects to an existing one if it does.
